rust/sdk/src/client.rs

```diff
@@ -388,6 +316,356 @@ pub(crate) fn to_lines(stream: impl Stream<Item = Result<Bytes, reqwest::Error>>
+                        let response = client.do_request(|c| c.post(publish_url).json(&request)).await?;
+                        let body = response.bytes().await?;
+                        Ok(serde_json::from_slice::<PublishResponse>(&body)?)
+                    };
+                    Publish::Pending(publish_response.boxed())
+                }
+                Publish::Pending(mut publish_response_future) => {
+                    let polled = publish_response_future.poll_unpin(cx);
+                    if polled.is_pending() {
+                        *this = Publish::Pending(publish_response_future);
+                    }
+                    return polled;
+                }
+                Publish::Empty => panic!("Polling a terminated Publish future"),
+            };
+        }
+    }
+}
+
+impl<'a> FusedFuture for Publish<'a> {
+    fn is_terminated(&self) -> bool {
+        if let Publish::Empty = self {
+            true
+        } else {
+            false
+        }
+    }
+}
+
+pub enum Query<'a> {
+    Initial {
+        client: &'a ActyxClient,
+        request: QueryRequest,
+    },
+    Pending(BoxFuture<'a, anyhow::Result<BoxStream<'a, QueryResponse>>>),
+    Empty,
+}
+
+impl<'a> Query<'a> {
+    fn new<Q: Into<String>>(client: &'a ActyxClient, query: Q) -> Self {
+        Self::Initial {
+            client,
+            request: QueryRequest {
+                query: query.into(),
+                lower_bound: Some(OffsetMap::empty()),
+                upper_bound: None,
+                order: Order::Asc,
+            },
+        }
+    }
+
+    /// Add a lower bound to the query.
+    ///
+    /// The lower bound limits the start of the query events.
+    /// As an example, consider the following (example) events:
+    /// ```json
+    /// { "lamport": 1, "event": { "temperature": 10 } }
+    /// { "lamport": 3, "event": { "temperature": 12 } }
+    /// { "lamport": 14, "event": { "temperature": 9 } }
+    /// ```
+    /// If you set the lower bound to `10`, only the last event will be returned.
+    pub fn with_lower_bound(mut self, lower_bound: OffsetMap) -> Self {
+        if let Self::Initial { ref mut request, .. } = self {
+            request.lower_bound = Some(lower_bound);
+        }
+        self
+    }
+
+    /// Add an upper bound to the query.
+    ///
+    /// The upper bound limits the start of the query events.
+    /// As an example, consider the following (example) events:
+    /// ```json
+    /// { "lamport": 1, "event": { "temperature": 10 } }
+    /// { "lamport": 3, "event": { "temperature": 12 } }
+    /// { "lamport": 14, "event": { "temperature": 9 } }
+    /// ```
+    /// If you set the upper bound to `10`, the first two events will be returned.
+    pub fn with_upper_bound(mut self, upper_bound: OffsetMap) -> Self {
+        if let Self::Initial { ref mut request, .. } = self {
+            request.upper_bound = Some(upper_bound);
+        }
+        self
+    }
```

Comment:
Since the default here means “present” and can thus not easily be represented using `.with_upper_bound()`, I think there should be a `.without_upper_bound()` companion.

Reply:
Should this be already implied by the non-calling of `with_upper_bound`?

Reply:
Ah, I see. I don't know that that pattern is a thing